### ISSUE TITLE
Still can't finish downloads 

### ISSUE THREAD

**Issue by Nickfost** [MultiPoolMiner/MultiPoolMiner#1229]
On issue #990 & #1008 i detailed the need for a feature for low bandwidth users. I'm here to once again ask for a better downloader, maybe even one that uses peer to peer technology. My miners have since moved from a 10mbps connection to a 600mbps connection. Not only have my miners not completed their downloads as you would expect, they are using hundreds of gigabytes of bandwidth.

**Comment by tbergeron** [comment]
I have to say, it would indeed be useful. Sometimes even on my very good connection I have trouble downloading binaries and I have to start all over again because for some reason it won't download the ones that error'd out on the first run.

**Comment by aaronsace** [comment]
They download fine on less then 10mbps; you must have problems with your connection. 

I'm not hosting the miners and they are coming straight from the creators own download pages...

I will have a look into it but don't be expecting me to create a magic downloader that is nothing to do with MultiPoolMiner.

**Comment by Nickfost** [comment]
I've tried a few connections recently. Tonight i'm going to swing by our miner colo location and reinstall with just the latest "official release" I'll report back about the success or failure of the downloads

**Comment by Nickfost** [comment]
On all of the miners i get a lot of "can not download" errors, i can confirm i am able to get a solid 20mbps to github/aws and 80mbps to everywhere else, i dont know if github imposes a rate limit or not and it doesnt seem anyone is talking about it anywhere i can find. 

suggestions, help, support, or anything similar would be nice. 


regarding p2p this _should_ be easily doable with a library like ipfs, no need to reinvent the wheel.

![h81bpkz 1](https://user-images.githubusercontent.com/1572453/36339098-7d4d096a-1383-11e8-9eb9-a0d064725ea7.png)

**Comment by UselessGuru** [comment]
> they are using hundreds of gigabytes of bandwidth.

I doubt this very much. 
A miner zip file is approx. 25MB each. Maybe something on your computer is blocking the unpacking process (Anti Virus quarantine?). Try again with deactivated AV. Most AV products block miner binaries.

**Comment by Nickfost** [comment]
There not an antivirus, this is six dedicated rigs in a colo facility. This bandwidth information is what the facility has reported to me with concern. I switched 1 rig to NHML to do a side by side comparison of bandwidth a few days ago. I’ll report back when I get that data. 

The colo facility is supposed to have VPN access where I’ll be able to remote into them (I wouldn't dare install TeamViewer) if you'd like we can arrange privately for you to remote into my rigs and take a look. 

The facility reports to me that all the concerning traffic is to AWS which I’m aware is where GitHub hosts.

**Comment by Nickfost** [comment]
> WARNING: Cannot download C:\Users\RDMiner2\Desktop\MPM\Bin\Lyra2RE2-NVIDIA\ccminer.exe distributed at https://github.com/nicehash/ccminer-nanashi/releases/download/1.7.6-r6/ccminer.zip.
> WARNING: Cannot find C:\Users\RDMiner2\Desktop\MPM\Bin\Lyra2RE2-NVIDIA\ccminer.exe distributed at https://github.com/nicehash/ccminer-nanashi/releases/download/1.7.6-r6/ccminer.zip.
> WARNING: Cannot download C:\Users\RDMiner2\Desktop\MPM\Bin\Lyra2z-NVIDIA\ccminer.exe distributed at https://github.com/djm34/ccminer-msvc2015/releases/download/v0.3.0/ccminer.rar.
> WARNING: Cannot find C:\Users\RDMiner2\Desktop\MPM\Bin\Lyra2z-NVIDIA\ccminer.exe
>  distributed at https://github.com/djm34/ccminer-msvc2015/releases/download/v0.3.0/ccminer.rar.
> WARNING: Cannot download C:\Users\RDMiner2\Desktop\MPM\Bin\NVIDIA-Nanashi\ccminer.exe distributed at https://github.com/Nanashi-Meiyo-Meijin/ccminer/releases/download/v2.2-mod-r2/2.2-mod-r2-CUDA9.binary.zip.
> WARNING: Cannot find C:\Users\RDMiner2\Desktop\MPM\Bin\NVIDIA-Nanashi\ccminer.exe distributed at https://github.com/Nanashi-Meiyo-Meijin/ccminer/releases/download/v2.2-mod-r2/2.2-mod-r2-CUDA9.binary.zip.
> WARNING: Cannot download C:\Users\RDMiner2\Desktop\MPM\Bin\Skunk-NVIDIA\ccminer.exe distributed at https://github.com/scaras/ccminer-2.2-mod-r1/releases/download/2.2-r1/2.2-mod-r1.zip.
> WARNING: Cannot find C:\Users\RDMiner2\Desktop\MPM\Bin\Skunk-NVIDIA\ccminer.exe
> distributed at https://github.com/scaras/ccminer-2.2-mod-r1/releases/download/2.2-r1/2.2-mod-r1.zip.
>

**Comment by Nickfost** [comment]
Regarding 
> WARNING: Cannot find C:\Users\RDMiner2\Desktop\MPM\Bin\Skunk-NVIDIA\ccminer.exe
> distributed at https://github.com/scaras/ccminer-2.2-mod-r1/releases/download/2.2-r1/2.2-mod-r1.zip.
> 

I believe the The file that's being looked for is available at this path.

> C:\Users\RDMiner2\Desktop\MPM\Bin\2.2-mod-r1\ccminer-2.2-mod-r1

It is indeed downloaded. is it not being extracted to the proper directory that's expected?

cc @aaronsace @UselessGuru 

Please advise

Thanks

**Comment by mkirov0** [comment]
Had a similar issue but after remembering to "Run as Administrator" the Start.bat file it worked like a charm! Could not be the case with your issue but just FYI.

**Comment by UselessGuru** [comment]
@Nickfost 
Why this fails in your case is hard to diagnose. 

As a **workaround** you can always manually unpack the miner binaries.
 
**Howto:** 
Look at the miner file
$URI tells you where to download the binaries from:
`$URI = "https://github.com/scaras/ccminer-2.2-mod-r1/releases/download/2.2-r1/2.2-mod-r1.zip"`

$Path describes where the miner binaries have to go:
`$Path = ".\Bin\Skunk-NVIDIA\ccminer.exe"`
Extract all files from the downloaded zip/rar file to the folder 'Bin\\\<folder name>' keeping the directory structure from the zip/rar file.

**Comment by mkirov0** [comment]
I tried exactly as @UselessGuru suggested above, but noticed most binaries were actually downloaded and/or extracted to ./bin folder.
In my case it was actually downloading them (except one which had the url changed), but could not unzip. At the same time MPM was giving me the "Cannot download" error. Some (ccminer instances mostly) were downloaded and unzipped in the ./bin folder, but could not run due to the MSVCR120.dll missing. Just FYI I did ALL the downloads mentioned in the documentation. Idk if it is a Windows 7 or a Microsoft issue, but had to download BOTH the 32-bit and 64-bit versions of MSVCR120.dll and install them one after the other to get it to work. Don't ask me how that fixed it, it just did 🤣

**Comment by Nickfost** [comment]
> $Path = ".\Bin\Skunk-NVIDIA\ccminer.exe"

the miners don't get extracted to there and it constantly re downloads.

**Comment by Nickfost** [comment]
> Had a similar issue but after making sure I have the MSVCR120.dll and remembering to "Run as Administrator" the Start.bat file it worked like a charm! Could not be the case with your issue but just FYI.

This is definitely not my issue. All of my accounts on all my machines are admin with uac off

**Comment by Nickfost** [comment]
I can confirm this issue on 7 different systems with two different internet connections.
I no longer think that this is a connection issue, it's a configuration issue. 
I believe the binaries are being extracted to an unexpected location and then MPM is thinking they aren't downloaded and redownloading them constantly (which would make sense for the hundreds and hundreds of gigabytes of bandwidth being used)

I’m using a clean installation the latest release. 

I cannot sustain this amount of wasted bandwidth as I plan to increase my rig’s count by 10-fold in the coming year.

I will say that not one person has offered to help or investigate my issue with any amount of seriousness which is disconcerting. I can offer remote access or full logs to help you diagnose my issue. 


Please advise.

**Comment by symalla** [comment]
I was helping a friend setup a rig at his place. I couldn't figure out why he was getting the download errors. In the end it turned out to be the DNS he was using that was blocking access to the sites as malicious to download the files. 

After switching him to googles DNS he was able complete the downloads successfully.

**Comment by Aspire1Inspire2** [comment]
I have this download issue almost for every miner. Downloading them from chrome or firefox using the same link is working. The downloader simply complains again and again about "cannot find" and "cannot download".

I tried this on windows 10 and windows server 2016 on two different networks. Both have defender and firewall disabled. Like previous posts, I don't think this is machine or network problem. I think its MPM configuration problem.

**Comment by Nickfost** [comment]
@Aspire1Inspire2 I've been hoping for a fix for three months. If you want to save on bandwidth you can delete downloader.ps1

it's a shame

**Comment by aaronsace** [comment]
It works on Windows 10 with a standard internet connection. 

PowerShell downloads them then 7Zip in extracts them; you can test this stuff yourselves. If you can't explain to me what you want me to change then there is not much I can do.

**Comment by Aspire1Inspire2** [comment]
> It works on Windows 10 with a standard internet connection.
@aaronsace 

I tested on a third Windows 10 PC in another network environment. The downloader is still not working. So far, I have tested three different physical locations, three different networks, all with windows defender and firewall disabled. All three don't work.

I tried firefox, chrome, and _wget_ in the Bash shell (Linux subsystem under Windows). All of them works and the correct zip file is downloaded to the same PC where the downloader complains "cannot find" and "cannot download".

I don't know much about PowerShell and the download mechanism. Is there any specific windows service for the PowerShell downloader that needs to be enabled? Is there any legacy command that needs to be updated from PowerShell 5 to 6? I don't know why it doesn't work, and the problem is reproducible for me.

Thank you for any of your suggestions.

**Comment by UselessGuru** [comment]
> Is there any specific windows service for the PowerShell downloader that needs to be enabled?

No. 

The actual code that does the download and unpacking is in 'include.psm1' in the function expand-WebRequest:

```
function Expand-WebRequest {
    [CmdletBinding()]
    param(
        [Parameter(Mandatory = $true)]
        [String]$Uri, 
        [Parameter(Mandatory = $false)]
        [String]$Path = ""
    )

    if (-not $Path) {$Path = Join-Path ".\Downloads" ([IO.FileInfo](Split-Path $Uri -Leaf)).BaseName}
    if (-not (Test-Path ".\Downloads")) {New-Item "Downloads" -ItemType "directory" | Out-Null}
    $FileName = Join-Path ".\Downloads" (Split-Path $Uri -Leaf)

    if (Test-Path $FileName) {Remove-Item $FileName}
    Invoke-WebRequest $Uri -OutFile $FileName -UseBasicParsing

    if (".msi", ".exe" -contains ([IO.FileInfo](Split-Path $Uri -Leaf)).Extension) {
        Start-Process $FileName "-qb" -Wait
    }
    else {
        $Path_Old = (Join-Path (Split-Path $Path) ([IO.FileInfo](Split-Path $Uri -Leaf)).BaseName)
        $Path_New = (Join-Path (Split-Path $Path) (Split-Path $Path -Leaf))

        if (Test-Path $Path_Old) {Remove-Item $Path_Old -Recurse}
        Start-Process "7z" "x `"$([IO.Path]::GetFullPath($FileName))`" -o`"$([IO.Path]::GetFullPath($Path_Old))`" -y -spe" -Wait

        if (Test-Path $Path_New) {Remove-Item $Path_New -Recurse}
        if (Get-ChildItem $Path_Old | Where-Object PSIsContainer -EQ $false) {
            Rename-Item $Path_Old (Split-Path $Path -Leaf)
        }
        else {
            Get-ChildItem $Path_Old | Where-Object PSIsContainer -EQ $true | ForEach-Object {Move-Item (Join-Path $Path_Old $_) $Path_New}
            Remove-Item $Path_Old
        }
    }
}
```

All that that function needs is the URL (where to dl from), and the path (where the miner binary must go)

I can not reproduce you issues. It is working fine for me.

**Comment by Aspire1Inspire2** [comment]
Thank you for the information to the specific download command, that is very helpful!!!
So here are some detailed debug information:

I ran this command:
`Invoke-WebRequest https://github.com/semtexzv/Prospector/releases/download/0.0.13-ALPHA/prospector-0.0.13-ALPHA-win64.zip -OutFile ./prospector-0.0.13-ALPHA-win64.zip -UseBasicParsing`

Here is what I get in my PowerShell (with Administrator):
`Invoke-WebRequest : The request was aborted: Could not create SSL/TLS secure channel.`
`At line:1 char:1`
`+ Invoke-WebRequest https://github.com/semtexzv/Prospector/releases/dow ...`
`+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`
`    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException`
`    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand`


Here is what I don't understand: Could not create SSL/TLS secure channel. Do I need to install OpenSSL or something like that?

**Comment by Aspire1Inspire2** [comment]
![image](https://user-images.githubusercontent.com/12798470/37043112-d98615b6-212d-11e8-8977-cf40a2d222bf.png)

@UselessGuru

**Comment by Aspire1Inspire2** [comment]
It seems that TLS 1.2 protocol needs to be enabled.
Any help would be really appreciated.

Thanks.

**Comment by Aspire1Inspire2** [comment]
[https://stackoverflow.com/questions/41618766/powershell-invoke-webrequest-fails-with-ssl-tls-secure-channel](url)

Here is what I did:
I add this line `[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12` befor the `Invoke-WebRequest $Uri -OutFile $FileName -UseBasicParsing`

Everything now works.

I created a pull request fixing the downloader problem.

Thank you.

@aaronsace @UselessGuru

**Comment by UselessGuru** [comment]
Thank you for your help. Hopefully this works for all DL problems!

**Comment by ScorpLAG123** [comment]
Where exactly did you add that line? @Aspire1Inspire2 I'm having miner download troubles as well...

**Comment by Nickfost** [comment]
Grabbed the latest zip on the releases page, deployed on the home (crap connection) test machine and will deploy on the 6 at the datacenter to see if the changes aid the issue

**Comment by Aspire1Inspire2** [comment]
@Nickfost
Not sure if the latest release in the zip file has thus change or not. You should check before you use.

**Comment by Nickfost** [comment]
Still wasting 100s of gigabytes of bandwidth....

**Comment by UselessGuru** [comment]
> Still wasting 100s of gigabytes of bandwidth....

You are definitively suffering from a problem specific to your environment. This is not a generic problem.
In your case downloading and unpacking all miners manually will do the trick. If you have multiple rigs you can even copy the 'Bin' folder without any side effect.

**Comment by Nickfost** [comment]
I don't know the paths i'd need to unpack too those instructions aren't very clear. 

Why not provide a large zipped bins folder in the releases page?

**Comment by Aspire1Inspire2** [comment]
I just would like to ask the question again, are you sure that in that
release you use, the change of the download related line is included?

Did you check? Or did you assume and proceed without check?

**Comment by UselessGuru** [comment]
> Why not provide a large zipped bins folder in the releases page?

Because we do not want to host binaries here. That's what the automatic DL is for - it will only download the miners that are really needed.

**Comment by Nickfost** [comment]
i downloaded [2.7.2.7](https://github.com/MultiPoolMiner/MultiPoolMiner/releases/download/v2.7.2.7/MultiPoolMinerV2.7.2.7.zip). I've got no idea if the change was included changelogs aren't updated on the releases page in months 

@UselessGuru if you provided the option to download a zip and extract them manually it could help people in my situation where the automatic downloads aren't working correctly. 


To recap, the downloader is downloading the files again and again and extracting them to seemingly wrong places and then it's assuming they aren't there so it's downloading them again. 

I'm unsure why there can't be a proper changelog or a bin download, it's quite annoying.

**Comment by Aspire1Inspire2** [comment]
It seems that it is just not there. I suggest you to go to the
github repository and download the "include.psm1" file and replace it.

Otherwise, you might be complaining about an issue that already get fixed
(though not in the released package).

**Comment by UselessGuru** [comment]
@Nickfost 

> why there can't be a proper changelog

Agreed. There is room for improvement, but there are not enough resources to document each and every feature. This is an evolving project. Anyone can read an compare the code, and there is a lot of info in the closed issues section (hey, you can search in github 👍 )

> or a bin download, it's quite annoying.

As said. We will not be hosting miner binaries that are NOT MPM territory. In the end we might even be held responsible for hosting malware.
You can find the DL URI and MP destination path info in each miner (as long as there is a direct DL link available that MPM can use).
You are suffering from a very rare problem and a workaround is available. Download and unpack manually. Do it once and copy the resulting Bin folder to your other rigs. Done.
